### PR TITLE
Move internal classes of the Nessie HTTP client

### PR DESCRIPTION
**Pull request by snazy**
The package `org.projectnessie.client.http` is pretty overloaded. It contains Nessie specific "public-ish" types, HTTP specific "public-ish" types and rather internal types. This change moves the internal types into a different package and also further abstracts the actual HTTP client implementation based on Java's `URLConnection` to a separate pacakge.

This change contains no functional change.

### CODE REVIEW

**Review by XN137 — APPROVED** [review]
generally looks ok but there seem to be some abstraction leaks, thus i am not sure if I understood the goal of the PR correctly :sweat_smile:

File: clients/client/src/main/java/org/projectnessie/client/http/impl/HttpRuntimeConfig.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.net.URI;
+import java.util.List;
+import javax.net.ssl.SSLContext;
+import org.immutables.value.Value;
+import org.projectnessie.client.http.RequestFilter;
+import org.projectnessie.client.http.ResponseFilter;
+
+/** Package-private HTTP configuration holder. */
```
[the remark this answers is not in the document]
Comment (snazy) [reply clients/client/src/main/java/org/projectnessie/client/http/impl/HttpRuntimeConfig.java:26]: oops

File: clients/client/src/main/java/org/projectnessie/client/http/HttpClient.java [hunk 2]
```diff
@@ -140,16 +128,20 @@ public HttpClient build() {
...
+              .mapper(mapper)
+              .readTimeoutMillis(readTimeoutMillis)
+              .connectionTimeoutMillis(connectionTimeoutMillis)
+              .isDisableCompression(disableCompression)
+              .sslContext(sslContext)
+              .addAllRequestFilters(requestFilters)
+              .addAllResponseFilters(responseFilters)
+              .build();
+
+      return new UrlConnectionClient(config);
```
[the remark this answers is not in the document]
Comment (snazy) [reply clients/client/src/main/java/org/projectnessie/client/http/HttpClient.java:144]: Yea - trying to get the new Java HTTP client (Java 11+) in.

File: clients/client/src/main/java/org/projectnessie/client/http/RequestContext.java [hunk 3]
```diff
@@ -21,6 +21,7 @@
 import java.util.Optional;
 import java.util.function.BiConsumer;
 import org.projectnessie.client.http.HttpClient.Method;
+import org.projectnessie.client.http.impl.HttpHeaders;
```
[the remark this answers is not in the document]
Comment (snazy) [reply clients/client/src/main/java/org/projectnessie/client/http/RequestContext.java:24]: good catch - will abstract that

File: clients/client/src/main/java/org/projectnessie/client/http/impl/RequestContextImpl.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  }
+
+  @Override
+  public Method getMethod() {
+    return method;
+  }
+
+  @Override
+  public Optional<Object> getBody() {
+    return body == null ? Optional.empty() : Optional.of(body);
```
Comment (XN137) [inline clients/client/src/main/java/org/projectnessie/client/http/impl/RequestContextImpl.java:77]: nit: do this once in constructor (ofNullable)

File: clients/client/src/test/java/org/projectnessie/client/auth/TestBasicAuthProvider.java [hunk 5]
```diff
@@ -126,7 +127,7 @@ void checkAuth(NessieAuthentication authentication) {
     assertThat(authFilter[0]).isInstanceOf(RequestFilter.class);
 
     HttpHeaders headers = new HttpHeaders();
-    RequestContext context = new RequestContext(headers, null, null, null);
+    RequestContext context = new RequestContextImpl(headers, null, null, null);
```
Comment (XN137) [inline clients/client/src/test/java/org/projectnessie/client/auth/TestBasicAuthProvider.java:130]: nit: looks odd to call a constructor with so many nulls (when usually method etc are not suppsoed to be null)...
maybe this and httpheaders above should be in some test utils as `newDummyRequestContext()` or the like (used in multiple places)

File: clients/client/src/main/java/org/projectnessie/client/http/impl/BaseHttpRequest.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+  protected OutputStream wrapOutputStream(OutputStream base) throws IOException {
+    return config.isDisableCompression()
+        ? new BufferedOutputStream(base)
+        : new GZIPOutputStream(base);
+  }
+
+  protected void writeBody(HttpRuntimeConfig config, OutputStream out, Object body)
+      throws IOException {
+    Class<?> bodyType = body.getClass();
+    if (bodyType != String.class) {
```
Comment (XN137) [inline clients/client/src/main/java/org/projectnessie/client/http/impl/BaseHttpRequest.java:72]: nit: == String and reverse if branches

**Review by XN137 — APPROVED** [review]

**Pull request merged**